File: sisrua_unified/py_engine/utils/schema_validator.py
"""
Lightweight JSON Schema draft-07 validator for py_engine.

Supports: type, properties, required, additionalProperties,
          minimum, maximum, exclusiveMinimum, exclusiveMaximum,
          minLength, maxLength, pattern, enum, items, minItems, maxItems,
          oneOf, anyOf, $ref (local #/definitions/…), propertyNames,
          const.

Falls back to jsonschema library when available for full spec coverage.
"""
from __future__ import annotations

import json
import os
import re
from typing import Any

# ── Optional: use jsonschema if available ────────────────────────────────────
try:
    import jsonschema  # type: ignore
    _JSONSCHEMA_AVAILABLE = True
except ImportError:
    _JSONSCHEMA_AVAILABLE = False

_SCHEMAS_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "schemas")


def _resolve_path(schema_path: str) -> str:
    if os.path.isabs(schema_path):
        return schema_path
    return os.path.normpath(os.path.join(_SCHEMAS_DIR, schema_path))


def _load_schema(schema_path: str) -> dict:
    resolved = _resolve_path(schema_path)
    with open(resolved, "r", encoding="utf-8") as fh:
        return json.load(fh)


def _get_definition(root: dict, ref: str) -> dict:
    if not ref.startswith("#/"):
        raise ValueError(f"Only local $ref supported, got: {ref}")
    parts = ref[2:].split("/")
    cur: Any = root
    for part in parts:
        cur = cur[part]
    return cur  # type: ignore[return-value]


def _python_type(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, int):
        return "integer"
    if isinstance(value, float):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "object"
    return type(value).__name__


def _validate_node(
    data: Any,
    schema: dict,
    root: dict,
    pointer: str,
    errors: list[str],
) -> None:
    # $ref
    if "$ref" in schema:
        resolved = _get_definition(root, schema["$ref"])
        _validate_node(data, resolved, root, pointer, errors)
        return

    # const
    if "const" in schema:
        if data != schema["const"]:
            errors.append(
                f"{pointer}: expected const {json.dumps(schema['const'])}, got {json.dumps(data)}"
            )
        return

    # oneOf
    if "oneOf" in schema:
        branches: list[dict] = schema["oneOf"]
        passing = [
            b for b in branches
            if not (sub := [], _validate_node(data, b, root, pointer, sub))[1] and not sub
        ]
        # simpler version without walrus for Python < 3.8
        passing_count = 0
        for branch in branches:
            sub: list[str] = []
            _validate_node(data, branch, root, pointer, sub)
            if not sub:
                passing_count += 1
        if passing_count != 1:
            errors.append(
                f"{pointer}: must match exactly one of {len(branches)} schemas (matched {passing_count})"
            )
        return

    # anyOf
    if "anyOf" in schema:
        branches = schema["anyOf"]
        any_pass = False
        for branch in branches:
            sub: list[str] = []
            _validate_node(data, branch, root, pointer, sub)
            if not sub:
                any_pass = True
                break
        if not any_pass:
            errors.append(
                f"{pointer}: must match at least one of {len(branches)} schemas"
            )
        return

    # type
    if "type" in schema:
        allowed_types = schema["type"] if isinstance(schema["type"], list) else [schema["type"]]
        actual = _python_type(data)
        # JSON Schema: integer is a subset of number
        matched = actual in allowed_types or (actual == "integer" and "number" in allowed_types)
        if not matched:
            errors.append(f"{pointer}: expected type {allowed_types}, got {actual}")
            return

    # enum
    if "enum" in schema:
        if data not in schema["enum"]:
            errors.append(
                f"{pointer}: must be one of {json.dumps(schema['enum'])}, got {json.dumps(data)}"
            )

    # string keywords
    if isinstance(data, str):
        if "minLength" in schema and len(data) < schema["minLength"]:
            errors.append(f"{pointer}: minLength {schema['minLength']}, got {len(data)}")
        if "maxLength" in schema and len(data) > schema["maxLength"]:
            errors.append(f"{pointer}: maxLength {schema['maxLength']}, got {len(data)}")
        if "pattern" in schema and not re.search(schema["pattern"], data):
            errors.append(f"{pointer}: does not match pattern {schema['pattern']}")

    # number keywords
    if isinstance(data, (int, float)) and not isinstance(data, bool):
        if "minimum" in schema and data < schema["minimum"]:
            errors.append(f"{pointer}: minimum {schema['minimum']}, got {data}")
        if "maximum" in schema and data > schema["maximum"]:
            errors.append(f"{pointer}: maximum {schema['maximum']}, got {data}")
        if "exclusiveMinimum" in schema and data <= schema["exclusiveMinimum"]:
            errors.append(f"{pointer}: exclusiveMinimum {schema['exclusiveMinimum']}, got {data}")
        if "exclusiveMaximum" in schema and data >= schema["exclusiveMaximum"]:
            errors.append(f"{pointer}: exclusiveMaximum {schema['exclusiveMaximum']}, got {data}")

    # array keywords
    if isinstance(data, list):
        if "minItems" in schema and len(data) < schema["minItems"]:
            errors.append(f"{pointer}: minItems {schema['minItems']}, got {len(data)}")
        if "maxItems" in schema and len(data) > schema["maxItems"]:
            errors.append(f"{pointer}: maxItems {schema['maxItems']}, got {len(data)}")
        if "items" in schema:
            item_schema = schema["items"]
            if isinstance(item_schema, list):
                for i, item in enumerate(data):
                    if i < len(item_schema):
                        _validate_node(item, item_schema[i], root, f"{pointer}[{i}]", errors)
            else:
                for i, item in enumerate(data):
                    _validate_node(item, item_schema, root, f"{pointer}[{i}]", errors)

    # object keywords
    if isinstance(data, dict):
        # propertyNames
        if "propertyNames" in schema:
            pn_schema = schema["propertyNames"]
            for key in data:
                _validate_node(key, pn_schema, root, f"{pointer}[key:{key}]", errors)

        # required
        for req in schema.get("required", []):
            if req not in data:
                errors.append(f"{pointer}: missing required property '{req}'")

        # properties
        props_schema: dict = schema.get("properties", {})
        for key, prop_schema in props_schema.items():
            if key in data:
                _validate_node(data[key], prop_schema, root, f"{pointer}.{key}", errors)

        # additionalProperties
        if schema.get("additionalProperties") is False and props_schema:
            known = set(props_schema.keys())
            for key in data:
                if key not in known:
                    errors.append(f"{pointer}: additional property '{key}' not allowed")


def validate_against_schema(
    data: dict,
    schema_path: str,
) -> tuple[bool, list[str]]:
    """Validate *data* against a JSON Schema v7 file.

    Args:
        data:        The dictionary to validate.
        schema_path: Absolute path or filename relative to schemas/ directory.

    Returns:
        ``(valid, errors)`` — *errors* is empty when *valid* is True.
    """
    errors: list[str] = []
    try:
        schema = _load_schema(schema_path)

        if _JSONSCHEMA_AVAILABLE:
            validator = jsonschema.Draft7Validator(schema)
            for err in validator.iter_errors(data):
                errors.append(f"{'/'.join(str(p) for p in err.absolute_path) or '#'}: {err.message}")
        else:
            _validate_node(data, schema, schema, "#", errors)

    except Exception as exc:  # noqa: BLE001
        errors.append(f"Schema load/parse error: {exc}")

    return (len(errors) == 0, errors)

File: sisrua_unified/server/utils/schemaValidator.ts
/**
 * Lightweight JSON Schema draft-07 validator.
 *
 * Supports: type, properties, required, additionalProperties,
 *           minimum, maximum, exclusiveMinimum, exclusiveMaximum,
 *           minLength, maxLength, pattern, enum, items, minItems, maxItems,
 *           oneOf, anyOf, $ref (local #/definitions/…), propertyNames,
 *           const, description (ignored).
 */
import fs from "fs";
import path from "path";

type SchemaDoc = Record<string, unknown>;

function resolvePath(schemaPath: string): string {
  if (path.isAbsolute(schemaPath)) return schemaPath;
  // Resolve relative to the project schemas/ directory
  const schemasDir = path.resolve(__dirname, "../../schemas");
  return path.join(schemasDir, schemaPath);
}

function loadSchema(schemaPath: string): SchemaDoc {
  const resolved = resolvePath(schemaPath);
  const raw = fs.readFileSync(resolved, "utf-8");
  return JSON.parse(raw) as SchemaDoc;
}

function getDefinition(root: SchemaDoc, ref: string): SchemaDoc {
  if (!ref.startsWith("#/"))
    throw new Error(`Only local $ref supported, got: ${ref}`);
  const parts = ref.slice(2).split("/");
  let cur: unknown = root;
  for (const part of parts) {
    if (typeof cur !== "object" || cur === null)
      throw new Error(`$ref path broken at: ${part}`);
    cur = (cur as SchemaDoc)[part];
  }
  return cur as SchemaDoc;
}

function validateNode(
  data: unknown,
  schema: SchemaDoc,
  root: SchemaDoc,
  pointer: string,
  errors: string[],
): void {
  // Resolve $ref
  if (typeof schema["$ref"] === "string") {
    const resolved = getDefinition(root, schema["$ref"]);
    validateNode(data, resolved, root, pointer, errors);
    return;
  }

  // const
  if ("const" in schema) {
    if (data !== schema["const"]) {
      errors.push(
        `${pointer}: expected const ${JSON.stringify(schema["const"])}, got ${JSON.stringify(data)}`,
      );
    }
    return;
  }

  // oneOf
  if (Array.isArray(schema["oneOf"])) {
    const branches = schema["oneOf"] as SchemaDoc[];
    const passing = branches.filter((branch) => {
      const sub: string[] = [];
      validateNode(data, branch, root, pointer, sub);
      return sub.length === 0;
    });
    if (passing.length !== 1) {
      errors.push(
        `${pointer}: must match exactly one of ${branches.length} schemas (matched ${passing.length})`,
      );
    }
    return;
  }

  // anyOf
  if (Array.isArray(schema["anyOf"])) {
    const branches = schema["anyOf"] as SchemaDoc[];
    const passing = branches.some((branch) => {
      const sub: string[] = [];
      validateNode(data, branch, root, pointer, sub);
      return sub.length === 0;
    });
    if (!passing) {
      errors.push(
        `${pointer}: must match at least one of ${branches.length} schemas`,
      );
    }
    return;
  }

  // type check
  if ("type" in schema) {
    const types = Array.isArray(schema["type"])
      ? schema["type"]
      : [schema["type"]];
    const actualType =
      data === null
        ? "null"
        : typeof data === "object"
          ? Array.isArray(data)
            ? "array"
            : "object"
          : typeof data;
    if (!types.includes(actualType)) {
      errors.push(
        `${pointer}: expected type ${types.join("|")}, got ${actualType}`,
      );
      return;
    }
  }

  // enum
  if (Array.isArray(schema["enum"])) {
    const allowed = schema["enum"] as unknown[];
    if (!allowed.includes(data)) {
      errors.push(
        `${pointer}: must be one of ${JSON.stringify(allowed)}, got ${JSON.stringify(data)}`,
      );
    }
  }

  // string keywords
  if (typeof data === "string") {
    if (
      typeof schema["minLength"] === "number" &&
      data.length < schema["minLength"]
    ) {
      errors.push(
        `${pointer}: minLength ${schema["minLength"]}, got ${data.length}`,
      );
    }
    if (
      typeof schema["maxLength"] === "number" &&
      data.length > schema["maxLength"]
    ) {
      errors.push(
        `${pointer}: maxLength ${schema["maxLength"]}, got ${data.length}`,
      );
    }
    if (
      typeof schema["pattern"] === "string" &&
      !new RegExp(schema["pattern"]).test(data)
    ) {
      errors.push(`${pointer}: does not match pattern ${schema["pattern"]}`);
    }
  }

  // number keywords
  if (typeof data === "number") {
    if (typeof schema["minimum"] === "number" && data < schema["minimum"]) {
      errors.push(`${pointer}: minimum ${schema["minimum"]}, got ${data}`);
    }
    if (typeof schema["maximum"] === "number" && data > schema["maximum"]) {
      errors.push(`${pointer}: maximum ${schema["maximum"]}, got ${data}`);
    }
    if (
      typeof schema["exclusiveMinimum"] === "number" &&
      data <= schema["exclusiveMinimum"]
    ) {
      errors.push(
        `${pointer}: exclusiveMinimum ${schema["exclusiveMinimum"]}, got ${data}`,
      );
    }
    if (
      typeof schema["exclusiveMaximum"] === "number" &&
      data >= schema["exclusiveMaximum"]
    ) {
      errors.push(
        `${pointer}: exclusiveMaximum ${schema["exclusiveMaximum"]}, got ${data}`,
      );
    }
  }

  // array keywords
  if (Array.isArray(data)) {
    if (
      typeof schema["minItems"] === "number" &&
      data.length < schema["minItems"]
    ) {
      errors.push(
        `${pointer}: minItems ${schema["minItems"]}, got ${data.length}`,
      );
    }
    if (
      typeof schema["maxItems"] === "number" &&
      data.length > schema["maxItems"]
    ) {
      errors.push(
        `${pointer}: maxItems ${schema["maxItems"]}, got ${data.length}`,
      );
    }
    if (schema["items"] !== undefined) {
      const itemSchema = schema["items"] as SchemaDoc | SchemaDoc[];
      if (Array.isArray(itemSchema)) {
        // tuple validation
        data.forEach((item, i) => {
          if (i < itemSchema.length) {
            validateNode(item, itemSchema[i], root, `${pointer}[${i}]`, errors);
          }
        });
      } else {
        data.forEach((item, i) => {
          validateNode(item, itemSchema, root, `${pointer}[${i}]`, errors);
        });
      }
    }
  }

  // object keywords
  if (typeof data === "object" && data !== null && !Array.isArray(data)) {
    const obj = data as Record<string, unknown>;

    // propertyNames
    if (
      typeof schema["propertyNames"] === "object" &&
      schema["propertyNames"] !== null
    ) {
      const pnSchema = schema["propertyNames"] as SchemaDoc;
      for (const key of Object.keys(obj)) {
        validateNode(key, pnSchema, root, `${pointer}[key:${key}]`, errors);
      }
    }

    // required
    if (Array.isArray(schema["required"])) {
      for (const req of schema["required"] as string[]) {
        if (!(req in obj)) {
          errors.push(`${pointer}: missing required property '${req}'`);
        }
      }
    }

    // properties
    const propsSchema = schema["properties"] as
      | Record<string, SchemaDoc>
      | undefined;
    if (propsSchema) {
      for (const [key, propSchema] of Object.entries(propsSchema)) {
        if (key in obj) {
          validateNode(obj[key], propSchema, root, `${pointer}.${key}`, errors);
        }
      }
    }

    // additionalProperties
    if (schema["additionalProperties"] === false && propsSchema) {
      const knownKeys = new Set(Object.keys(propsSchema));
      for (const key of Object.keys(obj)) {
        if (!knownKeys.has(key)) {
          errors.push(`${pointer}: additional property '${key}' not allowed`);
        }
      }
    }
  }
}

/**
 * Validate `data` against a JSON Schema v7 file.
 *
 * @param data       The value to validate.
 * @param schemaPath Absolute path or filename relative to schemas/ directory.
 * @returns          `{ valid, errors }` — errors is empty when valid.
 */
export function validateAgainstSchema(
  data: unknown,
  schemaPath: string,
): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  try {
    const schema = loadSchema(schemaPath);
    validateNode(data, schema, schema, "#", errors);
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    errors.push(`Schema load/parse error: ${msg}`);
  }
  return { valid: errors.length === 0, errors };
}
